[Question Answering] Adding Samples, Tests and Readme

AhmedLeithy opened the pull request:
This is a draft PR for adding tests and samples for the Question Answering. I will need to rebase this branch after #25348 gets merged, please ignore any changes related to the client itself for now.

What I've added so far is:
- Client Tests for validation. (Not all tests are activated yet because of [this issue](https://github.com/Azure/autorest.csharp/issues/1697))
- Setting up the `QuestionAnsweringProjectsClientLiveTestBase`.
- A sample with the title (sample3_CreateAndDeploy) as well as a markdown. This sample also includes adding a new knowledge base source, since we cannot deploy empty projects.
- A sample for Exporting and Importing Knowledge Bases (sample4_ExportAndImport)
- Adjusting and adding authoring methods to the migration guide
- Sample 5 with APIs for changing knowledge sources
- Adding the authoring client to the readme
- Add live tests in addition to the samples

What still needs to be done:
- Rebase this branch after [LLC initialization PR](https://github.com/Azure/azure-sdk-for-net/pull/25348) is merged.

AhmedLeithy committed:
Question Answering LLC Init

AhmedLeithy committed:
Question Answering LLC Init

AhmedLeithy committed:
Removed QuestionAnsweringProjectsClientOptions

AhmedLeithy committed:
Removed QuestionAnsweringProjectsClientOptions

AhmedLeithy committed:
Client Tests + First Sample Init

AhmedLeithy committed:
Fixed mistake with printed label.

azure-sdk commented:
API changes have been detected in `Azure.AI.Language.QuestionAnswering`. You can review API changes [here](https://apiview.dev/Assemblies/Review/83de30ac00694397980500611f43f9a7?diffOnly=True&diffRevisionId=ae1bde9350464af2a45eab9cc03c2857)

AhmedLeithy committed:
Sample 4 and migration guide initial commit

azure-sdk commented:
API changes have been detected in `Azure.AI.Language.QuestionAnswering`. You can review API changes [here](https://apiview.dev/Assemblies/Review/83de30ac00694397980500611f43f9a7?diffOnly=True&diffRevisionId=ae1bde9350464af2a45eab9cc03c2857)

AhmedLeithy committed:
Updated Migration Guide

AhmedLeithy committed:
Updated Delete LRO in sample

AhmedLeithy committed:
Added UpdateQnas test

AhmedLeithy committed:
Restructured TestBases and cleaned sample 4

AhmedLeithy committed:
Added export and import async

AhmedLeithy committed:
Added sample 4 markdown

AhmedLeithy committed:
Removing the delete section from sample 3

AhmedLeithy committed:
Fixed Typo

AhmedLeithy committed:
Added Sample5_UpdateKnowledgeSource

AhmedLeithy committed:
Added tests

AhmedLeithy committed:
Added addFeedback test

AhmedLeithy committed:
Added authoring client to ReadMe

AhmedLeithy committed:
Minor fixes after service update

AhmedLeithy committed:
Regenerated Client + Added reference to Azure.Core

AhmedLeithy committed:
Updated Tests and Samples with new API

AhmedLeithy committed:
Regenerated Snippets + Net standards
Removed QuestionAnsweringProjectsOptions

AhmedLeithy committed:
Set testing mode to Null

AhmedLeithy committed:
added TODO comment

AhmedLeithy committed:
Recorded Test Sessions

AhmedLeithy committed:
Re-exported netstandards

AhmedLeithy commented:
I have rebased this PR and Recorded all tests and samples. However, the tests that include the deploy() method are not yet functional due the the bug in the email thread.

AhmedLeithy commented:
On a seperate note, the CI pipeline seems to be complaining that some of the tests take too long to run (more than 10 seconds). Is there a way to adjust this 10 second value?

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/MigrationGuide.md:
```diff
@@ -9,9 +9,16 @@ Familiarity with the `Microsoft.Azure.CognitiveServices.Knowledge.QnAMaker` libr
-  - [Authenticating](#authenticating)
-  - [Querying a question](#querying-a-question)
-  - [Chatting](#chatting)
+  - [Runtime Client](#runtime-client)
+    - [Authenticating runtime client](#authenticating-runtime-client)
+    - [Querying a question](#querying-a-question)
+    - [Chatting](#chatting)
+  - [Authoring Client](#authoring-client)
+    - [Authenticating authoring client](#authenticating-authoring-client)
+    - [Creating knowledge base](#creating-knowledge-base)
+    - [Updating knowledge base](#updating-knowledge-base)
+    - [Exporting knowledge base](#exporting-knowledge-base)
+    - [Deleting knowledge base](#deleting-knowledge-base)
```
Be sure to check that all these exist / navigate to the bookmark. Seems like they should be correct but, for example, I think the automatic stubber doesn't do anything for words like "a" or "the", so line 14 may not navigate, for example.

AhmedLeithy replied:
Just checked them all, they are working (including line 14).

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/MigrationGuide.md:
````diff
@@ -91,3 +100,162 @@ AnswersOptions options = new AnswersOptions
+Now in `Azure.AI.Language.QuestionAnswering.Projects`, you create a `QuestionAnsweringProjectsClient` along with `AzureKeyCredential` from the package `Azure.Core`:
+
+```C# Snippet:Language_QnA_Maker_Snippets_MigrationGuide_CreateClient
+Uri endpoint = new Uri("https://myaccount.api.cognitive.microsoft.com");
````
We should use obvious placeholders like we do in existing samples e.g., `{questionanswering_uri}`.

AhmedLeithy replied:
Done

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/README.md:
````diff
@@ -119,6 +134,99 @@ foreach (KnowledgeBaseAnswer answer in response.Value.Answers)
+```C# Snippet:QuestionAnsweringProjectsClient_UpdateSourcesAsync
+// Set request content parameters for updating our new project's sources
+string sourceUri = "https://www.microsoft.com/en-in/software-download/faq";
+bool waitForCompletion = true;
````
Seems a little odd to add a local variable for this. If you want to pass `true` but also make it obvious what the parameter is for, consider using named parameters below where you pass this e.g., 

```csharp
await client.UpdateSourcesAsync(waitForCompletion: true, newProjectName, updateSourcesRequestContent);
```

AhmedLeithy replied:
Yea, this is a much better solution

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/README.md:
```diff
@@ -119,6 +134,99 @@ foreach (KnowledgeBaseAnswer answer in response.Value.Answers)
+Response creationResponse = await client.CreateProjectAsync(newProjectName, creationRequestContent);
+
+// Projects can be retrieved as follows
+AsyncPageable<BinaryData> projects = client.GetProjectsAsync();
```
As much as most of us would like to use async methods by default in samples, we tend to use sync methods since not everyone is familiar with async (especially `await foreach`). It's okay to have some async samples, but if doing so should have good reason and have a sync analog.

AhmedLeithy replied:
I've changed all the methods in the readme to be Sync. The rest of the samples have both sync and async versions.

However, for the migration guide I decided to keep our client's methods async for the following two reasons:
- The old cognitive service client only has Async methods (as far as I can tell), so it makes them to have them be the same type.
- The method calls are all one liners anyway. No unusual syntax like `await foreach`.

Does this seem okay with you?

heaths replied:
That's a great point and, yes, I agree it makes sense to use async in the migration guide. Thanks for considering that!

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/samples/Sample3_CreateAndDeploy.md:
````diff
@@ -0,0 +1,184 @@
+To create, deploy, or perform any other authoring actions for Question Answering projects, you need to first create a `QuestionAnsweringProjectsClient` using an endpoint and API key. These can be stored in an environment variable, configuration setting, or any way that works for your application.
+
+```C# Snippet:QuestionAnsweringProjectsClient_Create
+Uri endpoint = new Uri("https://myaccount.api.cognitive.microsoft.com");
````
As mentioned above (won't repeat, so please check all instances) let's use an obvious placeholder like you do below for the `{api-key}`.

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/samples/Sample4_ExportAndImport.md:
````diff
@@ -0,0 +1,121 @@
+
+```C# Snippet:QuestionAnsweringProjectsClient_ExportProject
+string exportFormat = "json"; // can also be tsv or excel
+bool waitForCompletion = true;
````
Same suggestion here (and check for all instances) about just using named params if you really want to emphasize what passing `true` would mean below. Just seems like an anti-pattern.

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/samples/Sample4_ExportAndImport.md:
````diff
@@ -0,0 +1,121 @@
+
+```C# Snippet:QuestionAnsweringProjectsClient_ImportProject
+// Set import project name and request content
+string importedProjectName = "importedProject";
+string importFormat = "json";
````
Nit: You could do the same here with named params as well.

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/tests/Infrastructure/QuestionAnsweringProjectsLiveTestBase.cs:
```diff
@@ -0,0 +1,79 @@
+
+        protected string CreateTestProjectName()
+        {
+            Recording.DisableIdReuse();
```
I'm not familiar with this, but perusing the code, I'm curious why you use it here? Of course we can't create the same IDs at the same time, but using the same ID *per test run* against test resources - as long as they get cleaned up - should be fine. Or is this only a problem because we're currently using static resources and old test resources may not have been cleaned up?

AhmedLeithy replied:
Yea I was considering the situation where an old test on a static resource may have not deleted the project it created. (I do call a delete project operation at the end of every test, but if the test fails, that part may not be executed).

I also recall using `Recording.DisableIdReuse()` in Translation when we were creating test blob containers and documents for translation. Although now that I think about it, that case was a bit different since we did not attempt to delete those containers after testing. 

Do you think we should keep it just in case?

heaths replied:
Talked to @pakrym and it only affects the recordings anyway - not the live tests - so probably worth removing. Sounds like it's not quite as effective now as it was originally designed to be anyway.

If we get conflicts on static live resources, it will be further incentive to get per-run live resources provisioned sooner than later.

AhmedLeithy replied:
Oh, I see. Removed

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/tests/QuestionAnsweringProjectsClientTests.cs:
```diff
@@ -0,0 +1,321 @@
+
+        public QuestionAnsweringProjectsClient Client => new(Endpoint, new AzureKeyCredential("5d757e79a27f40f896d1dc0c04fb3d09"));
+
+        // private String SampleProjectName = "sampleProjectName";
+
+        // private String SampleDeploymentName = "sampleDeploymentName";
```
Nit: I'd just remove dead code.

AhmedLeithy replied:
These variables will be referenced by the tests below once they are activated, so I decided to temporarily comment them out. I can remove them and bring them back once the generation bug is fixed if that is better.

heaths replied:
Naw, that's fine to leave them then. Just make sure to use `string` not `String`.

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/tests/QuestionAnsweringProjectsClientTests.cs:
```diff
@@ -0,0 +1,321 @@
+            Assert.AreEqual("projectName", ex.ParamName);
+        }
+
+        // These tests should be activated once the bug with the validation generation is fixed.
```
Could you create a tracking bug and update the comment with the link as well? That's what we typically do so these TODO (or TODO-like comments: also recommend adding a "TODO" at the beginning since some tools like Visual Studio will actually find and display those in a task list) comments to make sure they aren't lost.

AhmedLeithy replied:
Done! I also added TODOs and links for the other issues that we are tracking (transforms and .Core project reference)

heaths reviewed (changes requested):
Looks great, and while I have most nits (more like suggestions), samples should generally be all sync instead of async.

AhmedLeithy committed:
Added TODOs with links for pre-release changes

AhmedLeithy committed:
Updated samples to address PR comments

AhmedLeithy committed:
Added clear string value for import project name

AhmedLeithy committed:
changing uri to clear URI

AhmedLeithy committed:
Fixed some missed variable names in sample 3

AhmedLeithy committed:
added project name declaration for clarity

AhmedLeithy committed:
Resolving final conversations

AhmedLeithy committed:
Restructured readme Examples headers to avoid CI error

AhmedLeithy committed:
Fixing async method call

AhmedLeithy committed:
Fixes for slow tests

heaths committed:
Merge branch 'main' into QA-LLC-SamplesTestsAndReadme

heaths commented:
@AhmedLeithy I removed the ProjectReference to Azure.Core since that shipped - or is the new [1.22.0 version](https://www.nuget.org/packages/Azure.Core/1.22.0) not the one you were expecting? - but looks like tests might need to be re-recorded, not to mention the test duration issue we discussed Thursday on Teams is happening.

AhmedLeithy committed:
Sync WaitForCompletion work around

AhmedLeithy committed:
Merge branch 'QA-LLC-SamplesTestsAndReadme' of https://github.com/AhmedLeithy/azure-sdk-for-net into QA-LLC-SamplesTestsAndReadme

AhmedLeithy committed:
Disabled deployment tests

AhmedLeithy committed:
Removed InstrumentOperation used for testing

AhmedLeithy committed:
Added workaround to !SNIPPET condition

AhmedLeithy commented:
@heaths thanks for handling the project reference/package issue. This PR now closes #26171. 

I fixed the issue we discussed in teams regarding the slowed tests. I also disabled tests that include the `deployProject()` method for now, so that we can merge this PR. I added a [tracking issue](https://github.com/Azure/azure-sdk-for-net/issues/26401) to remind us to renable those tests once the bug is fixed.

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/tests/Samples/Sample5_UpdateKnowledgeSources.cs:
```diff
@@ -46,8 +46,12 @@ public void KnowledgeSources()
             updateSourcesOperation.WaitForCompletion();
+#else
+            await updateSourcesOperation.WaitForCompletionAsync();
+#endif
```
You can't just keep this method sync and call the sync version of `WaitForCompletion()`? If not necessary, I'd just keep it simple / easy to maintain and easier to understand if a user views the original sample source (this).

AhmedLeithy replied:
Unfortunately this runs into the timeout issue when running the test in playback. Using the async version of `WaitForCompletionAsync()` was part of the fix Christopher suggested. We can change this back once the bug is fixed.

AhmedLeithy replied:
Oh and btw, I would have merged this PR, but I do not have the write access to do that.

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/tests/Samples/Sample5_UpdateKnowledgeSources.cs:
```diff
@@ -56,12 +60,12 @@ public void KnowledgeSources()
                 Console.WriteLine(source);
             }
-            #endregion
+#endregion
```
Nit: we typically align the `#region`s with the source, but it ultimately doesn't matter (just a bit easier to read, IMO).

AhmedLeithy committed:
Aligned regions with source
